pkg/apis/pipeline/v1beta1/taskrun_validation.go

```diff
@@ -114,27 +114,64 @@ func (ts *TaskRunSpec) validateInlineParameters(ctx context.Context) (errs *apis
 	if ts.TaskSpec == nil {
 		return errs
 	}
-	var paramSpec []ParamSpec
+	paramSpecForValidation := make(map[string]ParamSpec)
 	for _, p := range ts.Params {
 		pSpec := ParamSpec{
 			Name:    p.Name,
 			Default: &p.Value,
+			Type:    p.Value.Type,
+		}
+		if p.Value.ObjectVal != nil {
+			pSpec.Properties = make(map[string]PropertySpec)
+			prop := make(map[string]PropertySpec)
+			for k := range p.Value.ObjectVal {
+				prop[k] = PropertySpec{Type: ParamTypeString}
+			}
+			pSpec.Properties = prop
 		}
-		paramSpec = append(paramSpec, pSpec)
+		paramSpecForValidation[p.Name] = pSpec
 	}
 	for _, p := range ts.TaskSpec.Params {
-		skip := false
-		for _, ps := range paramSpec {
-			if ps.Name == p.Name {
-				skip = true
-				break
+		if _, ok := paramSpecForValidation[p.Name]; ok {
+			if p.Default != nil && p.Default.ObjectVal != nil {
+				for k, v := range p.Default.ObjectVal {
+					if paramSpecForValidation[p.Name].Default.ObjectVal == nil {
+						paramSpecForValidation[p.Name].Default.ObjectVal = map[string]string{k: v}
+					} else {
+						paramSpecForValidation[p.Name].Default.ObjectVal[k] = v
+					}
+					if paramSpecForValidation[p.Name].Properties == nil {
+						if pSpec, ok := paramSpecForValidation[p.Name]; ok {
+							pSpec.Properties = map[string]PropertySpec{
+								k: {Type: ParamTypeString},
+							}
+							paramSpecForValidation[p.Name] = pSpec
+						}
+					} else {
+						if pSpec, ok := paramSpecForValidation[p.Name]; ok {
+							pSpec.Properties[k] = PropertySpec{Type: ParamTypeString}
+							paramSpecForValidation[p.Name] = pSpec
+						}
+					}
+				}
 			}
+		} else {
+			if p.Default != nil && p.Default.ObjectVal != nil {
+				prop := make(map[string]PropertySpec)
+				for k := range p.Default.ObjectVal {
+					prop[k] = PropertySpec{Type: ParamTypeString}
+				}
+				p.Properties = prop
+			}
+			paramSpecForValidation[p.Name] = p
```

Comment:
Feel like this part is a bit hard to read. How about decomposing them into helper function that does the merge job? 

Two benefits I can think of: 1. easy to understand, 2. it can be reused in `pipelinerun_validation.go` as well.
```
=========
                var the_map map[string]string
		if v, ok := paramSpecForValidation[p.Name]; ok {
			the_map = mergeObjectPartOfParamValue(v.Default, p.Default)
		} else {
			the_map = mergeParamValueObjectPart(nil, p.Default)
		}
		paramSpecForValidation[p.Name].Default = the_map
		paramSpecForValidation[p.Name].Properties = composePropertySectionFromAMap(the_map)
		
==========	
helpers
==========
		
func mergeObjectPartOfParamValue(input1 ParamValue, input2 ParamValue) map[string]string {
             // check nil of input1 and input2
             // if same key exist in both, value in input1 will take precedence 
}
		
func composePropertySectionFromAMap(input map[string]string) map[string]PropertySpec {
}
```

Reply:
Agreed. I was about to do that after I got all the tests passing.

Reply:
Done now.